In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import sys
from datetime import datetime, timedelta
import re
import sys

# SQL and Database Test Instructions
- The top 18 columns show a simplistic overview of 3 tables we have at Overstock.com.  
- Your task will be to write SQL as if you were quering these tables to retreive the data as specified in the questions.
- The questions will require you to join tables and make calculations within the SQL to get the desired output.
- Please note column E (notes) to verify you are using columns and data correctly as you write the SQL.

| Database 	| Tables 	| Columns 	| Value type 	| Notes 	|
|------------------	|-------------------------	|-------------------------------	|------------	|------------------------------------------------------------------------------------------------------------	|
| EDW_ACCESS_VIEWS 	| BI_VISIT 	| VISIT_ID 	| DECIMAL 	| Primary Unique Key - ID for individual browsing sessions on website 	|
| EDW_ACCESS_VIEWS 	| BI_VISIT 	| VISITOR_ID 	| DECIMAL 	| ID for individuals on website 	|
| EDW_ACCESS_VIEWS 	| BI_VISIT 	| VISIT_DT 	| DATE 	| Date of visit to website 	|
| EDW_ACCESS_VIEWS 	| BI_VISIT 	| FIRST_PAGE_VIEW_CAMPAIGN_ID 	| DECIMAL 	| Campaign ID for the first webpage a visit views.  The campaign that drove the visitor to view the website. 	|
| EDW_ACCESS_VIEWS 	| BI_VISIT 	| PAGE_VIEW_COUNT 	| INTEGER 	| A count of pages viewed in a given visit 	|
| EDW_ACCESS_VIEWS 	| BI_SALES 	| SALES_TANSACTION_ID 	| INTEGER 	| ID for individual sale transactions 	|
| EDW_ACCESS_VIEWS 	| BI_SALES 	| SALES_TRANSACTION_LINE_ID 	| INTEGER 	| Primary Unique Key - ID for individual items purchased in a sale transaction 	|
| EDW_ACCESS_VIEWS 	| BI_SALES 	| SALES_TAANSACTION_CAMPAIGN_ID 	| DECIMAL 	| The last campaign a visit clicked through before the given transaction 	|
| EDW_ACCESS_VIEWS 	| BI_SALES 	| REVENUE_TOP_LINE 	| DECIMAL 	| Top line revenue for a transaction 	|
| EDW_ACCESS_VIEWS 	| BI_SALES 	| REVENUE_BOTTOM_LINE 	| DECIMAL 	| Bottom line revenue for a transaction 	|
| EDW_ACCESS_VIEWS 	| BI_SALES 	| VISIT_ID 	| DECIMAL 	| ID for individual browsing sessions on website 	|
| EDW_ACCESS_VIEWS 	| BI_SALES 	| PRODUCT_ID 	| DECIMAL 	| ID for item purchased 	|
| EDW_ACCESS_VIEWS 	| BI_SALES 	| DEPARTMENT 	| SMALLINT 	| Department of the product in a given transaction 	|
| EDW_ACCESS_VIEWS 	| BI_SALES 	| SALES_TRANSACTION_DATE 	| DATE 	| Date of sale 	|
| EDW_ACCESS_VIEWS 	| BI_CAMPAIGN_DESCRIPTION 	| CAMPAIGN_ID 	| DECIMAL 	| Primary Unique Key - Campaign ID for each marketing campaign run 	|
| EDW_ACCESS_VIEWS 	| BI_CAMPAIGN_DESCRIPTION 	| CAMPAIGN_DESCRIPTION 	| VARCHAR 	| Description of the campaign 	|
| EDW_ACCESS_VIEWS 	| BI_CAMPAIGN_DESCRIPTION 	| CHANNEL_DESCRIPTION 	| VARCHAR 	| Marketing channel the campaign belongs to 	|

### SQL and Database Test - Question 1
- Show me sale revenue (top line & bottom line) by Marketing channel for the past 90 days trended by day.

In [4]:
"""
SELECT
sales.SALES_TRANSACTION_DATE,
campaigns.CHANNEL_DESCRIPTION,
SUM(sales.REVENUE_TOP_LINE) AS revenue_top_line_sum,
SUM(sales.REVENUE_BOTTOM_LINE) AS revenue_bottom_line_sum
FROM EDW_ACCESS_VIEWS.BI_SALES AS sales
LEFT JOIN EDW_ACCESS_VIEWS.BI_CAMPAIGN_DESCRIPTION AS campaigns ON (campaigns.CAMPAIGN_ID = visits.SALES_TAANSACTION_CAMPAIGN_ID)
WHERE VISIT_DT BETWEEN DATE(DATE_SUB(NOW(), INTERVAL 90 DAY)) AND DATE(NOW())
GROUP BY 1,2
ORDER BY 1 DESC
"""

### SQL and Database Test - Question 2
- Show me the number of visits, visitors and bounce rate for each Marketing Channel for the past 90 days trended by day. (Hint: Bounce rate = Single page view visits / total visits)

In [ ]:
"""
SELECT
visits.VISIT_DT,
campaigns.CHANNEL_DESCRIPTION,
COUNT(*) AS visit_count,
COUNT(DISTINCT(visits.VISITOR_ID)) AS visitor_count,
SUM(COALESCE((CASE WHEN visits.PAGE_VIEW_COUNT = 1 THEN 1 ELSE 0 END), 0)) / COUNT(*) AS bounce_rate
FROM EDW_ACCESS_VIEWS.BI_VISIT AS visits
LEFT JOIN EDW_ACCESS_VIEWS.BI_CAMPAIGN_DESCRIPTION AS campaigns ON (campaigns.CAMPAIGN_ID = visits.FIRST_PAGE_VIEW_CAMPAIGN_ID)
WHERE VISIT_DT BETWEEN DATE(DATE_SUB(NOW(), INTERVAL 90 DAY)) AND DATE(NOW())
GROUP BY 1,2
ORDER BY 1 DESC
"""

### SQL and Database Test - Question 3
- Show me the number of visits each Marketing channel acquired and the revenue (top line & Bottom line) for each last clicked Marketing channel for the past 90 days trended by day.  Calculate conversion rate (orders/visits). Answer in one query.

In [ ]:
"""
SELECT
COUNT(VISIT_ID),
COUNT(VISITOR_ID)
FROM EDW_ACCESS_VIEWS.BI_VISIT
WHERE VISIT_DT BETWEEN DATE(DATE_SUB(NOW(), INTERVAL 90 DAY)) AND DATE(NOW())
"""

# Basic Functions 1
Please show your work (i.e., show your formulas)

- Use data/basic-functions-1.csv
- Using Excel's built in functionality, split the values in column D into first and last names

In [5]:
bf1_df = pd.read_csv('data/basic-functions-1.csv', header=None, names=["full_name"])
bf1_df['last_name'], bf1_df['first_name'] = zip(*bf1_df['full_name'].map(lambda x: x.split(', ')))
bf1_df

full_name    last_name first_name
0            Aitken, Don       Aitken        Don
1      Anderson, Melanie     Anderson    Melanie
2         Attridge, Matt     Attridge       Matt
3        Barnes, Melissa       Barnes    Melissa
4             Bates, Jen        Bates        Jen
5         Beausheur, Kim    Beausheur        Kim
6             Bey, April          Bey      April
7        Blakeslee, Adam    Blakeslee       Adam
8       Boutsis, Vanessa      Boutsis    Vanessa
9          Bown, Heather         Bown    Heather
10    Broadhead, Annette    Broadhead    Annette
11     Broadhead, Candis    Broadhead     Candis
12         Brown, Ashlee        Brown     Ashlee
13       Bruss, Michella        Bruss   Michella
14          Buhler, Cami       Buhler       Cami
15      Burgess, Heather      Burgess    Heather
16        Callahan, Lois     Callahan       Lois
17         Carter, Craig       Carter      Craig
18  Christensen, Jessica  Christensen    Jessica
19           Clark, Cami        Clark       Cami
20          Clark, Shana        Clark      Shana
21        Crown, Merissa        Crown    Merissa
22           Decker, Sam       Decker        Sam
23      DeHart, Danielle       DeHart   Danielle
24      Dimick, Tichelle       Dimick   Tichelle
25         Dunn, Valerie         Dunn    Valerie
26          Dyer, Kendra         Dyer     Kendra
27      Elangovan, Sunny    Elangovan      Sunny
28         Ellison, Todd      Ellison       Todd

# Basic Functions 2
- Use basic-functions-2.csv
- Using a formula in cell G5, count how many associates have quality greater than 90% and pieces per hour greater than 45
- Using a formula in cell G9, what is the average quality and pieces per hour for the associates on Jason's team?
- Using a formula in cell G13, what is the average pieces per hour for associates with quality lower than 89%?

In [7]:
bf2_df = pd.read_csv('data/basic-functions-2.csv')
bf2_df.columns = ["associate", "team", "quality", "pieces_per_hour"] 
bf2_df

associate    team  quality  pieces_per_hour
0   Michella  Stacie     0.97               48
1     Ashlee   Jason     0.90               48
2    Annette     Kim     0.98               40
3     Kendra     Kim     0.80               44
4    Heather  Stacie     0.84               45
5       Cami  Stacie     0.86               39
6       Adam   Jason     0.88               40
7      Sunny  Stacie     0.83               45
8       Cami     Kim     0.91               46
9        Sam  Stacie     1.00               40
10      Matt  Stacie     0.83               45
11    Candis     Kim     0.89               38
12     Clint  Stacie     0.97               42
13  Tichelle  Stacie     0.89               43
14   Heather     Kim     0.88               44
15   Melanie   Jason     0.88               45
16       Kim     Kim     0.95               47
17   Jessica   Jason     0.90               47
18   Melissa   Jason     0.98               39
19       Don     Kim     0.99               44
20   Kristin  Stacie     0.88               48
21     Craig  Stacie     0.97               38
22   Merissa  Stacie     0.91               42
23      Jade   Jason     0.87               40
24      Todd     Kim     0.95               45
25  Danielle   Jason     0.89               39
26      Gile   Jason     0.87               44
27      Lois   Jason     0.84               48
28     April   Jason     0.81               40
29    Bonnie   Jason     0.83               40
30     Shana     Kim     0.85               41
31   Tiffany   Jason     0.98               39
32       Jen   Jason     0.95               41
33   Vanessa  Stacie     0.92               43
34   Valerie     Kim     0.99               39

### Basic Functions 2 - Question 1
- Count how many associates have quality greater than 90% and pieces per hour greater than 45

In [8]:
bf2_df[(bf2_df.quality > 0.90) & (bf2_df.pieces_per_hour > 45)]

associate    team  quality  pieces_per_hour
0   Michella  Stacie     0.97               48
8       Cami     Kim     0.91               46
16       Kim     Kim     0.95               47

### Basic Functions 2 - Question 2
- What is the average quality and pieces per hour for the associates on Jason's team?

In [9]:
jasons_team_df = bf2_df[(bf2_df.team == "Jason")]
jasons_team_df

associate   team  quality  pieces_per_hour
1     Ashlee  Jason     0.90               48
6       Adam  Jason     0.88               40
15   Melanie  Jason     0.88               45
17   Jessica  Jason     0.90               47
18   Melissa  Jason     0.98               39
23      Jade  Jason     0.87               40
25  Danielle  Jason     0.89               39
26      Gile  Jason     0.87               44
27      Lois  Jason     0.84               48
28     April  Jason     0.81               40
29    Bonnie  Jason     0.83               40
31   Tiffany  Jason     0.98               39
32       Jen  Jason     0.95               41

In [10]:
jasons_team_df.quality.describe()

count    13.000000
mean      0.890769
std       0.052830
min       0.810000
25%       0.870000
50%       0.880000
75%       0.900000
max       0.980000
Name: quality, dtype: float64

In [11]:
jasons_team_df.pieces_per_hour.describe()

count    13.000000
mean     42.307692
std       3.568020
min      39.000000
25%      40.000000
50%      40.000000
75%      45.000000
max      48.000000
Name: pieces_per_hour, dtype: float64

### Basic Functions 2 - Question 3
- Use basic-functions-3.csv
- Using Excel's built in functionality fill in the Quality column.
    - DO NOT USE VLOOKUP or change order of Associate names.

In [12]:
bf3_a_df  = pd.read_csv('data/basic-functions-3-a.csv')
bf3_a_df.columns = ["associate", "associate_id", "quality"]
bf3_a_df = bf3_a_df.drop(['quality'], axis=1)
bf3_a_df

associate  associate_id
0   Michella          6466
1     Ashlee          4936
2    Annette          4749
3     Kendra          7134
4    Heather          1426
5       Cami            59
6       Adam          8222
7      Sunny          7176
8        Sam          3716
9       Matt          4908
10    Candis          8813
11     Clint          6902
12  Tichelle          8013
13   Melanie          5399
14       Kim          7597
15   Jessica          7399
16   Melissa          7250
17       Don          8735
18   Kristin          5885
19     Craig          5575
20   Merissa          1799
21      Jade          4823
22      Todd          6320
23  Danielle          8058
24      Gile          2491
25      Lois           175
26     April          7301
27    Bonnie          1887
28     Shana          8500
29   Tiffany          5859
30       Jen           951
31   Vanessa          7462
32   Valerie          6956

In [13]:
bf3_b_df = pd.read_csv('data/basic-functions-3-b.csv')
bf3_b_df.columns = ["associate_id", "quality"]
bf3_b_df

associate_id  quality
0             59     0.86
1            175     0.89
2            951     0.85
3           1426     0.84
4           1799     0.88
5           1887     0.84
6           2491     0.95
7           3716     0.91
8           4749     0.98
9           4823     0.97
10          4908     1.00
11          4936     0.90
12          5399     0.89
13          5575     0.99
14          5859     0.83
15          5885     0.98
16          6320     0.91
17          6466     0.97
18          6902     0.89
19          6956     0.95
20          7134     0.80
21          7176     0.83
22          7250     0.95
23          7301     0.87
24          7399     0.88
25          7462     0.98
26          7597     0.88
27          8013     0.97
28          8058     0.87
29          8222     0.88
30          8500     0.81
31          8735     0.90
32          8813     0.83

In [14]:
joined_table = pd.merge(bf3_a_df, bf3_b_df, how='left', left_on=["associate_id"], right_on=["associate_id"])
joined_table

associate  associate_id  quality
0   Michella          6466     0.97
1     Ashlee          4936     0.90
2    Annette          4749     0.98
3     Kendra          7134     0.80
4    Heather          1426     0.84
5       Cami            59     0.86
6       Adam          8222     0.88
7      Sunny          7176     0.83
8        Sam          3716     0.91
9       Matt          4908     1.00
10    Candis          8813     0.83
11     Clint          6902     0.89
12  Tichelle          8013     0.97
13   Melanie          5399     0.89
14       Kim          7597     0.88
15   Jessica          7399     0.88
16   Melissa          7250     0.95
17       Don          8735     0.90
18   Kristin          5885     0.98
19     Craig          5575     0.99
20   Merissa          1799     0.88
21      Jade          4823     0.97
22      Todd          6320     0.91
23  Danielle          8058     0.87
24      Gile          2491     0.95
25      Lois           175     0.89
26     April          7301     0.87
27    Bonnie          1887     0.84
28     Shana          8500     0.81
29   Tiffany          5859     0.83
30       Jen           951     0.85
31   Vanessa          7462     0.98
32   Valerie          6956     0.95

# Paid Digital Marketing Analysis

- You are currently in position 3 of a paid text keyword search.  
    - The click-through rate (clicks per impressions) is 1% and the cost-per-click CPC is $1.

- You can get into position 1 by paying $2 CPC and would expect a CTR of 2%.

- In position 2, your CPC would be $1.5 and you would expect a CTR of 1.5%.												
- Conversion Rate (conversions per click) is 5% for all positions. (CvR)												
- Gross margin per conversion is 20% (before advertising cost).												

- Average Order (conversion) Value is $200. (AOV)	

The table below summarizes the above information:

| Ad Position | CTR   | CvR  | Gross Margin | CPC     | AOV       |
|-------------|-------|------|--------------|---------|-----------|
| 1           | 0.02  | 0.05 | 0.2          | \\$2.00 | \\$200.00 |
| 2           | 0.015 | 0.05 | 0.2          | \\$1.50 | \\$200.00 |
| 3           | 0.01  | 0.05 | 0.2          | \\$1.00 | \\$200.00 |

### Paid Digital Marketing Analysis - Questions

1. What is the cost per conversion in position 1?
2. What is the cost per conversion in position 2?
3. What is the cost per conversion in position 3?
4. If goal is to maximize absolute net profit, then which position you would want to be in?
5. Compare that with the most profitable position (from the answer of the previous question) and calculate the % extra cost you would have to pay to sustain this position?
6. What is return on ad spend (ROAS) of the additional conversions gained by moving from position 3 to position 2?

# Email Analysis

**Instruction:** Please read the below and conduct an analysis on the email marketing data provided.						
Email lead generation vendors (leadgen) provide the email marketing team with leads(email addresses). You plan to use the below dataset to help a business manager make decisions on which leads to buy.						
This data consists of leads from two separate vendors. The sales and engagement data spans the 30 days after we received the leads.						

- **acct_id** - An ID representing the customer to which the email was sent						
- **email_grp** - A grouping of email addresses representing different leadgen vendors						
- **sales** - What the customer payed when purchasing products						
- **net_profit** - Overstock's actual profit on the customer's purchase						
- **send** - Number of email sent to the customer						
- **opens** - Number of emails the customer opened						
- **unsubscribed** - Indicate that the customer unsubscribed from future Overstock email sends						

In [15]:
ea_df  = pd.read_csv('data/email-analysis.csv')
ea_df

acct_id email_grp  sales    net_profit  sends  opens  unsubscribed
0      138,473,050      Grp2      0             0     30      0             0
1      138,473,051      Grp2      0             0     46      0             0
2      138,473,089      Grp2      0             0     29      0             0
3      138,473,093      Grp2      0             0     49      5             0
4      138,473,094      Grp2      0             0     46      2             0
5      138,473,095      Grp2      0             0     31      0             0
6      138,473,112      Grp2      0             0     48      2             0
7      138,473,131      Grp1      0             0     56     48             0
8      138,473,133      Grp2      0             0     49      1             0
9      138,473,134      Grp2      0             0     29      0             0
10     138,473,135      Grp2      0             0     50     20             0
11     138,473,136      Grp2      0             0     49     39             0
12     138,473,144      Grp2      0             0     29      0             0
13     138,473,145      Grp2      0             0      1      0             0
14     138,473,160      Grp2      0             0     45      9             0
15     138,473,161      Grp2      0             0     47      1             0
16     138,473,179      Grp2      0             0     29      0             0
17     138,473,197      Grp2      0             0     30      0             0
18     138,473,207      Grp2      0             0     31      0             0
19     138,473,209      Grp1      0             0     44      0             0
20     138,473,214      Grp1      0             0     36      5             1
21     138,473,216      Grp2      0             0     29      0             0
22     138,473,217      Grp2      0             0     52      5             0
23     138,473,218      Grp2      0             0     30      0             0
24     138,473,228      Grp1      0             0     47      3             0
25     138,473,231      Grp2      0             0     45      1             0
26     138,473,235      Grp2      0             0     46      1             0
27     138,473,239      Grp1      0             0     45      0             0
28     138,473,252      Grp2      0             0     46     15             0
29     138,473,253      Grp2      0             0     30      0             0
...            ...       ...    ...           ...    ...    ...           ...
28163  138,741,089      Grp1      0             0      9      4             1
28164  138,741,093      Grp1      0             0     12      0             0
28165  138,741,116      Grp1      0             0     37     40             0
28166  138,741,132      Grp1      0             0      7      3             1
28167  138,741,161      Grp1      0             0     16      0             0
28168  138,741,201      Grp1      0             0     17      0             0
28169  138,741,232      Grp1      0             0     33      2             0
28170  138,741,266      Grp1  65.99  -8.207585895     49      9             0
28171  138,741,270      Grp1      0             0     17      0             0
28172  138,741,294      Grp1      0             0     35      1             0
28173  138,741,300      Grp1      0             0     17      0             0
28174  138,741,359      Grp1      0             0     17      0             0
28175  138,741,394      Grp1      0             0     31      0             0
28176  138,741,396      Grp1      0             0     34      2             0
28177  138,741,397      Grp1      0             0     16      0             0
28178  138,741,402      Grp1      0             0     36     11             0
28179  138,741,426      Grp1      0             0     16      0             0
28180  138,741,446      Grp1      0             0     17      0             0
28181  138,741,479      Grp1      0             0     31      0             0
28182  138,741,488      Grp1     

### Email Analysis - Question 1
- Please evaluate and analyze the leads for each group.

### Email Analysis - Question 2
- What recommendations to the business manager would you make? 

### Email Analysis - Question 3
- Present your analysis and explain your thought process/logic/assumptions when making your recommendations. 

# Ad Placement Case Analysis

### Instructions
We run ads on a 3rd party websites that give the customers a coupon for product on our website.
For a holiday promotion last year, we ran a 10% off coupon and paid $7,000 for an ad placement on their website.
We have decided to run a 12% (increased from 10% last year) coupon and pay for the same placement this year and expect to receive a 15% year over year lift in revenue.

### Ad Placement Case Analysis - Question 1 
- Knowing the information provided above, fill in This Year estimates using the data provided below.

|              | Last Year Numbers | This Year's Estimates |
|--------------|-------------------|-----------------------|
| Revenue      | 850,000.00        |                       |
| Margin       | 155,000.00        |                       |
| Coupon Cost  | 85,000.00         |                       |
| Ad Spend     | 7,000.00          |                       |
| Gross Profit | 63,000.00         |                       |

### Ad Placement Case Analysis - Question 2

- The advertiser wants Overstock.com to pay for an additional one time ad placement on their website for an additional \\$7,000 cost.
- Using This Year's Estimates you calculated in the questions above, determine the revenue lift from this additional placement we would need to break even.

|              | This Year's Estimates |
|--------------|-----------------------|
| Revenue      |                       |
| Margin       |                       |
| Gross Profit |                       |
| Coupon Cost  |                       |
| Ad Spend     |                       |

### Ad Placement Case Analysis - Question 3
- Assume we will get no lift in revenue from the additional \\$7,000.00 cost placement, what changes to our coupon could we make to still break even?